SeQuant/core/math.hpp

```diff
@@ -15,30 +16,42 @@ constexpr std::size_t pow2(std::size_t n) {
   return 1ul << n;
 }
 
-constexpr inline std::size_t factorial(std::size_t n) {
-  assert(n >= 0 && n <= 20);
-  const std::size_t values[] = {1,
-                                1,
-                                2,
-                                6,
-                                24,
-                                120,
-                                720,
-                                5040,
-                                40320,
-                                362880,
-                                3628800,
-                                39916800,
-                                479001600,
-                                6227020800,
-                                87178291200,
-                                1307674368000,
-                                20922789888000,
-                                355687428096000,
-                                6402373705728000,
-                                121645100408832000,
-                                2432902008176640000};
-  return values[n];
+using int128_t = boost::multiprecision::int128_t;
+constexpr inline int128_t factorial(std::size_t n) {
+  assert(n >= 0);
+  const int128_t values[] = {1,
+                             1,
+                             2,
+                             6,
+                             24,
+                             120,
+                             720,
+                             5040,
+                             40320,
+                             362880,
+                             3628800,
+                             39916800,
+                             479001600,
+                             6227020800,
+                             87178291200,
+                             1307674368000,
+                             20922789888000,
+                             355687428096000,
+                             6402373705728000,
+                             121645100408832000,
+                             2432902008176640000};
+  if (n <= 20)
+    return values[n];
+  else {
+    auto fac = [](std::size_t n) {
+      int128_t result = 1;
+      for (auto i = 2; i <= n; ++i) {
+        result *= i;
+      }
+      return result;
+    };
+    return fac(n);
```

Comment:
1. Why define a lamda for calculating the factorial and then immediately calling that instead of simply computing the factorial straight away
2. You don't need to recompute the factorial up to `n==20`:
```suggestion
  int128_t result = values[20];
  for (auto i = 21; i <= n; ++i) {
    result *= i;
  }
```
or even better:
```suggestion
  constexpr auto nPrecomputed = std::size(values);
  int128_t result = values[nPrecomputed - 1];
  for (auto i = nPrecomputed + 1; i <= n; ++i) {
    result *= i;
  }
  return result;
```

Reply:
thanks, addressed, `fac` is gone